src/Features/Core/Portable/Snippets/RoslynLSPSnippetConverter.cs

```diff
@@ -0,0 +1,161 @@
+            return ConvertToLSPSnippetString(extendedTextChange, placeholders, caretPosition);
+        }
+
+        /// <summary>
+        /// Iterates through every index in the snippet string and determines where the
+        /// LSP formatted chunks should be inserted for each placeholder.
+        /// </summary>
+        private static string ConvertToLSPSnippetString(TextChange textChange, ImmutableArray<SnippetPlaceholder> placeholders, int caretPosition)
+        {
+            var textChangeStart = textChange.Span.Start;
+            var textChangeText = textChange.NewText;
+            Contract.ThrowIfNull(textChangeText);
+
+            using var _ = PooledStringBuilder.GetInstance(out var lspSnippetString);
+            using var disposer = PooledDictionary<int, (string identifier, int priority)>.GetInstance(out var dictionary);
+            GetMapOfSpanStartsToLSPStringItem(ref dictionary, placeholders, textChangeStart);
+
+            // Need to go through the length + 1 since caret postions occur before and after the
+            // character position.
+            // If there is a caret at the end of the line, then it's position
+            // will be equivalent to the length of the TextChange.
+            for (var i = 0; i < textChangeText.Length + 1;)
+            {
+                if (i == caretPosition - textChangeStart)
+                {
+                    lspSnippetString.Append("$0");
+
+                    // Special case for cursor position since they will occur between positions
+                    // so we still wants to insert the character following the cursor position.
+                    // Will not happen for placeholders since they have a direct mapping from position
+                    // of the identifier to their position in the TextChange text.
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                    }
+
+                    i++;
+                }
+
+                //Tries to see if a value exists at that position in the map, and if so it
+                // generates a string that is LSP formatted.
+                if (dictionary.TryGetValue(i, out var placeholderInfo))
+                {
+                    var str = $"${{{placeholderInfo.priority}:{placeholderInfo.identifier}}}";
+                    lspSnippetString.Append(str);
+
+                    // Skip past the entire identifier in the TextChange text
+                    i += placeholderInfo.identifier.Length;
+                }
+                else
+                {
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                        i++;
+                    }
+                    else
+                    {
+                        break;
```

Comment:
this break feels superfluous.

Reply:
it would be an infinite loop otherwise in this case:

`$$ [|if ({|placeholder:true|})
{
}|]`

Because we get stuck never incrementing the iterator. I could also just increment at this point and the loop ends, but that would be equivalent to breaking?